src/main/kotlin/org/opensearch/replication/task/shard/BatchSizeSettings.kt

```diff
@@ -0,0 +1,84 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.replication.task.shard
+
+import org.opensearch.index.IndexSettings
+import org.opensearch.replication.ReplicationPlugin
+import org.opensearch.replication.ReplicationSettings
+
+/**
+ * Helper class to manage batch size settings with fallback from index-level to cluster-level
+ */
+class BatchSizeSettings(
+    private val indexSettings: IndexSettings,
+    private val replicationSettings: ReplicationSettings
+) {
+
+    /**
+     * Get the effective batch size - index-level if set, otherwise cluster-level
+     */
+    fun getBatchSize(): Int {
+        return if (hasIndexLevelSetting()) {
+            ReplicationPlugin.REPLICATION_FOLLOWER_OPS_BATCH_SIZE_INDEX.get(indexSettings.settings)
+        } else {
+            replicationSettings.batchSize
+        }
+    }
+
+    /**
+     * Check if index-level setting is configured
+     */
+    fun hasIndexLevelSetting(): Boolean {
+        return indexSettings.settings.hasValue(ReplicationPlugin.REPLICATION_FOLLOWER_OPS_BATCH_SIZE_INDEX.key)
+    }
+
+    /**
+     * Get the source of the current batch size setting
+     */
+    fun getBatchSizeSource(): String {
+        return if (hasIndexLevelSetting()) "index-level" else "cluster-level"
+    }
+
+    // For dynamic batch size adjustment (2GB fix)
+    @Volatile
+    private var dynamicBatchSize: Int? = null
+    private val minBatchSize = 16
+
+    /**
+     * Get effective batch size considering dynamic adjustments
+     */
+    fun getEffectiveBatchSize(): Int {
+        return dynamicBatchSize ?: getBatchSize()
+    }
+
+    /**
+     * Reduce batch size for 2GB limit handling
+     */
+    fun reduceBatchSize() {
+        val currentSize = getEffectiveBatchSize()
```

Comment:
The reduced batch size value is in memory. It'll get cleaned up after the node restart or shard moves. Let's also manually test the behavior is such cases